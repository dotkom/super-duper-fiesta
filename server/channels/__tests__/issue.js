jest.mock('../../models/issue');
jest.mock('../../utils');
const { createIssue, closeIssue, adminDeleteIssue } = require('../issue');
const { addIssue, endIssue, deleteIssue } = require('../../models/issue');
const { emit, broadcast } = require('../../utils');
const { generateSocket, generateIssue } = require('../../utils/generateTestData');

beforeEach(() => {
  addIssue.mockImplementation(async () => generateIssue());
  endIssue.mockImplementation(async () => generateIssue());
});

describe('createIssue', () => {
  const generateData = () => ({});
  it('emits OPEN_ISSUE action creates issue', async () => {
    await createIssue(generateSocket(), generateData());

    expect(emit.mock.calls).toMatchSnapshot();
    expect(broadcast.mock.calls).toMatchSnapshot();
  });

  it('emits an error when it fails', async () => {
    addIssue.mockImplementation(async () => { throw new Error('Failed'); });
    await createIssue(generateSocket(), generateData());

    expect(emit.mock.calls).toMatchSnapshot();
    expect(broadcast.mock.calls).toEqual([]);
  });
});

describe('closeIssue', () => {
  const generateData = () => ({
    issue: generateIssue(),
  });

  it('emits close issue action and disables voting', async () => {
    endIssue.mockImplementation(async () => generateIssue({ active: false }));
    await closeIssue(generateSocket(), generateData());

    expect(emit.mock.calls).toMatchSnapshot();
    expect(broadcast.mock.calls).toMatchSnapshot();
  });

  it('emits error when it fails', async () => {
    endIssue.mockImplementation(async () => { throw new Error('Failed'); });
    await closeIssue(generateSocket(), generateData());

    expect(emit.mock.calls).toMatchSnapshot();
    expect(broadcast.mock.calls).toEqual([]);
  });
});

describe('adminDeleteIssue', () => {
  const generateData = () => ({
    issue: generateIssue(),
  });
  it('emits delete issue on success', async () => {
    deleteIssue.mockImplementation(async () => generateIssue({ active: false, deleted: true }));

    await adminDeleteIssue(generateSocket(), generateData());

    expect(emit.mock.calls).toMatchSnapshot();
    expect(broadcast.mock.calls).toMatchSnapshot();
  });
});
