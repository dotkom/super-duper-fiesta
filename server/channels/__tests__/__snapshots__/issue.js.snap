// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adminDeleteIssue emits delete issue on success 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "CLOSE_ISSUE",
    Object {
      "_id": 1,
      "active": false,
      "deleted": true,
      "description": "Description goes here",
      "genfors": "1",
    },
  ],
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "DELETED_ISSUE",
    Object {
      "_id": 1,
      "active": false,
      "deleted": true,
      "description": "Description goes here",
      "genfors": "1",
    },
  ],
]
`;

exports[`adminDeleteIssue emits delete issue on success 2`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "DISABLE_VOTING",
  ],
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "CLOSE_ISSUE",
    Object {
      "_id": 1,
      "active": false,
      "deleted": true,
      "description": "Description goes here",
      "genfors": "1",
    },
  ],
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "DELETED_ISSUE",
    Object {
      "_id": 1,
      "active": false,
      "deleted": true,
      "description": "Description goes here",
      "genfors": "1",
    },
  ],
]
`;

exports[`closeIssue emits close issue action and disables voting 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "CLOSE_ISSUE",
    Object {
      "_id": 1,
      "active": false,
      "description": "Description goes here",
      "genfors": "1",
    },
  ],
]
`;

exports[`closeIssue emits close issue action and disables voting 2`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "DISABLE_VOTING",
  ],
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "CLOSE_ISSUE",
    Object {
      "_id": 1,
      "active": false,
      "description": "Description goes here",
      "genfors": "1",
    },
  ],
]
`;

exports[`closeIssue emits error when it fails 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "issue",
    Object {},
    Object {
      "error": "Closing issue failed",
    },
  ],
]
`;

exports[`createIssue emits OPEN_ISSUE action creates issue 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "OPEN_ISSUE",
    Object {
      "_id": 1,
      "active": true,
      "description": "Description goes here",
      "genfors": "1",
    },
    Object {
      "action": "open",
    },
  ],
]
`;

exports[`createIssue emits OPEN_ISSUE action creates issue 2`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "OPEN_ISSUE",
    Object {
      "_id": 1,
      "active": true,
      "description": "Description goes here",
      "genfors": "1",
    },
    Object {
      "action": "open",
    },
  ],
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "ENABLE_VOTING",
  ],
]
`;

exports[`createIssue emits an error when it fails 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "issue",
    Object {},
    Object {
      "error": "Adding new question failed",
    },
  ],
]
`;
